# Part D - Increase the number of hidden layers

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

In [1]:
#import libraries
import pandas as pd
import numpy as np
import keras

from keras.models import Sequential
from keras.layers import Dense, Input

from sklearn.model_selection import train_test_split

import warnings
warnings.simplefilter('ignore', FutureWarning)

2025-01-21 18:12:54.227024: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1737479574.251662  556517 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1737479574.261202  556517 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-01-21 18:12:54.283708: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


### Import and check data

In [2]:
# download the data and read it into a pandas dataframe.
concrete_data = pd.read_csv("concrete_data.csv")

concrete_data.shape

(1030, 9)

In [3]:
# Already done in part A
#concrete_data.head()
#concrete_data.tail()
#concrete_data.isnull().sum()

#### Split data into predictors and target

In [4]:
predictors = concrete_data[concrete_data.columns[concrete_data.columns != 'Strength']] 
target = concrete_data['Strength'] 
print(f"predictors shape: {predictors.shape}")
print(f"target shape: {target.shape}")

predictors shape: (1030, 8)
target shape: (1030,)


#### Normalize the data

In [5]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()

### Build the neural network

In [6]:
# define regression model with 3 hiden layers now
def regression_model():
    # create model
    model = Sequential()
    model.add(Input(shape=(8,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')   
    return model

### Repeat 50 times and show mean and standard deviation of the results

In [7]:
model = regression_model()
mean_squared_errors = []

2025-01-21 18:12:57.469252: E external/local_xla/xla/stream_executor/cuda/cuda_driver.cc:152] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


In [8]:
%%time

for i in range(50):
    # i) split data
    x_train, x_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=9)
    ## ii) fit model
    model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=50, verbose=0)
    # iii) evaluate model
    mse = model.evaluate(x_test, y_test, verbose=0)
    mean_squared_errors.append(mse)

CPU times: user 6min 30s, sys: 21.5 s, total: 6min 51s
Wall time: 5min 42s


Calculate mean and std. deviation

In [9]:
#Calculate the mean and standard deviation of the mean squared errors
mse_mean = np.mean(mean_squared_errors)
mse_std = np.std(mean_squared_errors)

print(f"Mean of calculated mse: {mse_mean}")
print(f"Standard deviation of calculated mse: {mse_std}")

Mean of calculated mse: 56.93883010864258
Standard deviation of calculated mse: 10.916908658774547


How does the mean of the mean squared errors compare to that from Step B?

First execution:<br>
Mean of calculated mse: 54.81726654052734 <br>
Std. deviation of calculated mse: 8.387985485964666  <br>
CPU times: user 6min 23s, sys: 21.6 s, total: 6min 44s <br>
Wall time: 5min 33s
<br>
<br>
Second execution:<br>
Mean of calculated mse:  56.93883010864258  <br>
Std. deviation of calculated mse:  10.916908658774547 <br>
CPU times: user 6min 30s, sys: 21.5 s, total: 6min 51s <br>
Wall time: 5min 42s

There is a .. quite appreciable improvement comparing with part B. (Specially regarding std. deviation). <br>
The results here are also similar to those in part C with epoch=100, but it seems that the additional hidden layers achieve the same result in much less time.


<br>